### PR TITLE
Issue #3951: SeparatorWrap: add support for method reference operator

### PR DESCRIPTION
**Pull request by Luolc**
#3951 

Firstly, we need to confirm the check logic. As the issue mentioned: `want '::' operator at end of previous line`, but it is not correct according to google style: http://checkstyle.sourceforge.net/reports/google-java-style-20170228.html#s4.5.1-line-wrapping-where-to-break

> When a line is broken at a non-assignment operator the break comes before the symbol.
> - This also applies to the following "operator-like" symbols:
>   - the two colons of a method reference (::)

Notice that the break should come before `::`. 

Configurations for Google style is updated as well.

~~Diff report is on the way.~~

Diff report:
http://www.luolc.com/checkstyle-diff-report/issue3951/

### CODE REVIEW

**Comment by Luolc** [comment]
Got an error when generating diff report:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-site-plugin:3.3:site (default-site) on project sample: Error during page generation: Error rendering Maven report: Failed during checkstyle configuration: cannot initialize module TreeWalker - Token "METHOD_REF" was not found in Acceptable tokens list in check com.puppycrawl.tools.checkstyle.checks.whitespace.SeparatorWrapCheck
```

I did add `METHOD_REF` into Acceptable tokens list in check in the commit: https://github.com/checkstyle/checkstyle/pull/4149/files#diff-9e446cf94e6a0626e011eff85209b645R135

Maybe it is because in the master branch `METHOD_REF` is not in Acceptable tokens.

$ cat my_check.xml
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">


<module name = "Checker">
    <property name="charset" value="UTF-8"/>
    
    <!-- do not change severity to 'error', as that will hide errors caused by exceptions -->
    <property name="severity" value="warning"/>

    <!-- haltOnException is required for exception fixes and reporting of all exceptions -->
    <property name="haltOnException" value="false"/>

    <!-- BeforeExecutionFileFilters is required for sources of java9 -->
    <module name="BeforeExecutionExclusionFileFilter">
        <property name="fileNamePattern" value="module\-info\.java$" />
    </module>

    <module name="TreeWalker">
         <!-- Example of sevntu.checkstyle Check usage -->
         <!-- <module name="NestedSwitchCheck"/> -->
         
         <!-- Example of checkstyle Check usage -->
         <module name="SeparatorWrap">
            <property name="tokens" value="METHOD_REF" />
            <property name="option" value="nl" />
        </module>
    </module>
<!--
    <module name="SeverityMatchFilter">
        <property name="severity" value="warning"/>
        <property name="acceptOnMatch" value="false"/>
    </module>
-->
</module>
```

**Comment by MEZk** [comment]
@Luolc 
As METHOD_REF caused confusion, please add an example into xdoc which clearly shows where the violation should be and where it should not.

**Comment by Luolc** [comment]
@MEZk 
> please add an example into xdoc which clearly shows where the violation should be and where it should not.

Done. Please have a check whether the example is ok.

> Please, update your checkstyle-tester local repo

Done. New diff report is generated. Posted at same location: http://www.luolc.com/checkstyle-diff-report/issue3951/

**Comment by MEZk** [comment]
@Luolc 
[This change](https://github.com/checkstyle/contribution/commit/7275e9ca89630cf92ddf724b84619d1d7ae57441#diff-043b0f45d71a6070eb9ae93696bbcdd7L251) is not in your local repo.

File: src/xdocs/config_whitespace.xml [hunk 1]
```diff
@@ -1630,6 +1632,18 @@ foo(i,
     s);
         </source>
         <p>
+          Example for method reference (good case and bad case):
+        </p>
+        <source>
+// violation, breaking after method reference is bad practice
```
Comment (MEZk) [inline src/xdocs/config_whitespace.xml]: Please, put the comment on the line where the violation is actually happens.

**Comment by Luolc** [comment]
@MEZk Oh... my fault. I only did rebase on my master branch. Sorry for that.
New report is on the way.

File: src/xdocs/config_whitespace.xml (hunk 1 above)
Comment (Luolc) [inline src/xdocs/config_whitespace.xml]: Done.

**Comment by Luolc** [comment]
@MEZk new report: http://www.luolc.com/checkstyle-diff-report/issue3951/
What does the warning means? Did I miss smth?

**Comment by MEZk** [comment]
@Luolc 
> What does the warning means? Did I miss smth?

It means that diff.groovy does not take into account any excludes specified in projects-to-test-on.properties. It runs the report generation against all source files of the target projects.

File: src/xdocs/config_whitespace.xml [hunk 2]
```diff
@@ -1630,6 +1632,15 @@ foo(i,
     s);
         </source>
         <p>
+          Example for method reference (good case and bad case):
+        </p>
+        <source>
+Arrays.sort(stringArray, String:: // violation, breaking after method reference is bad practice
```
Comment (MEZk) [inline src/xdocs/config_whitespace.xml]: It is not a general bad practice. User is able to decide whether he/she wants to configure the check to rise the violation if ```::``` is on the next line after line break. Please remove ```breaking after method reference is bad practice```.
Please, add an example of check's configuration as METHOD_REF is not in the set of default tokens.

Comment (Luolc) [inline src/xdocs/config_whitespace.xml]: Done.

**Comment by romani** [comment]
> Firstly, we need to confirm the check logic. As the issue mentioned: want '::' operator at end of previous line, but it is not correct according to google style:

Styles are different from team to team, Check as ability to satisfy both of them by property `option`
http://checkstyle.sourceforge.net/property_types.html#wrapOp , http://checkstyle.sourceforge.net/config_whitespace.html#SeparatorWrap .

> Should Google Style Guide be updated in the scope of this PR?

I did not find issue, looks like we miss it. @Luolc , please create new issue for this fix, after merge of this PR.

**Comment by Luolc** [comment]
@romani 
> please split this commit into two

Done.

> please generate reports on projects that have "::" in mode `eol` to see violations

Done. New report with `EOL` option: http://www.luolc.com/checkstyle-diff-report/issue3951/index.html
Still no regression.

One of Travis-CI's job failed. Would you please restart it? https://travis-ci.org/checkstyle/checkstyle/jobs/220246004

Error log of the failed job:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test (default-test) on project checkstyle: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:2.19.1:test failed: The forked VM terminated without properly saying goodbye. VM crash or System.exit called?
```

**Comment by rnveach** [comment]
> One of Travis-CI's job failed.

restarted

**Comment by romani** [comment]
> Still no regression.

that is weird that with both options values there is no regression it could mean only that "::" is not used in this projects or there is not wraps ..... we need to see some violations on real code.

Please run Check with both values against https://github.com/javaslang/javaslang  , if you can find any other project that have wrapping of "::" it will be awesome.

**Comment by rnveach** [comment]
@romani @Luolc 
I ran report to show projects that are using `::` and where if you wish to verify why any aren't causing differences.
http://rveach.no-ip.org/checkstyle/regression/reports/17/

**Comment by Luolc** [comment]
@romani 
http://www.luolc.com/checkstyle-diff-report/issue3951-nl/
http://www.luolc.com/checkstyle-diff-report/issue3951-eol/
Besides javaslang, I also tried groovy and groovy-eclipse, the last two are the projects which @eric-milles are working on, the issue opener.

Unfortunately there are still no regression. :(

@rnveach I see wrapping of `::` in checkstyle's code, but they are in the input files. Does diff report generator process them?

**Comment by rnveach** [comment]
> Does diff report generator process them?

@Luolc Yes, this is why we added the red warning label to the diff reports to remind us we process all files in the project.

**Comment by rnveach** [comment]
http://www.luolc.com/checkstyle-diff-report/issue3951-nl/groovy/index.html

@Luolc How did you run regression with tokens `METHOD_REF`?
It is a new token and not in master. I got the following exception when trying to run regression on it.

> Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Token "METHOD_REF" was not found in Acceptable tokens list in check com.puppycrawl.tools.checkstyle.checks.whitespace.SeparatorWrapCheck

Since this is a brand new token, you can only run `launch.groovy` on the change, not `diff.groovy`.

**Comment by Luolc** [comment]
@rnveach I found the error at first. To generate diff report, in master branch I add `METHOD_REF` to Acceptable tokens set (without changing other functional codes in the check).

**Comment by Luolc** [comment]
@rnveach @romani 
The way I generated diff reports is not reasonable at all...

Since it is a new token I could only use `launch.groovy`.

http://www.luolc.com/checkstyle-diff-report/reports/issue3951/nl/
http://www.luolc.com/checkstyle-diff-report/reports/issue3951/eol/

These are the reports, we finally find `::` with wrapping, but only in checkstyle, our project itself. And it is only in input files.

BTW, how could I generate reports with root page like http://rveach.no-ip.org/checkstyle/regression/reports/17/ ?
I run `groovy launch.groovy -l projects-to-test-on.properties -c my_check.xml` but only get separate folders without a root `index.html`. The pages above is written manually.

**Comment by rnveach** [comment]
> BTW, how could I generate reports with root page like http://rveach.no-ip.org/checkstyle/regression/reports/17/ ?

@Luolc My system is my own creation. It was made before the groovy scripts which try to copy most of my functionality with added support of non-Linux OSes.
It is located at https://github.com/checkstyle/contribution/blob/master/checkstyle-tester/launch_diff.sh but it isn't really recommended for regression but I will help you in gitter if you want to try it. I can do various things in mine which groovy doesn't support yet.

I will make a new issue to have groovy do regression with no base for new checks/tokens.

**Comment by romani** [comment]
> Firstly, we need to confirm the check logic. As the issue mentioned: want '::' operator at end of previous line, but it is not correct according to google style

In issue I did not found any, reference to google style, so issue was general and correct (violation was expected as config is required this)
But thanks a lot for update of Google config and its ITs - we missed this. Even http://checkstyle.sourceforge.net/google_style.html shows full coverage of for this point of style.

I am ok merge.
I am fine with reports. We tried enough to find regression. If some problem appear they will not be very obvious , so community will not blame us for simple regression.

@rnveach , please do final review and merge.

**Pull request merged**